Python/ceval.c

```diff
@@ -4223,6 +4225,65 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, InterpreterFrame *frame, int thr
             DISPATCH();
         }
 
+        case TARGET(LOAD_METHOD_ADAPTIVE): {
+            assert(cframe.use_tracing == 0);
+            SpecializedCacheEntry *cache = GET_CACHE();
+            if (cache->adaptive.counter == 0) {
+                PyObject *owner = TOP();
+                PyObject *name = GETITEM(names, cache->adaptive.original_oparg);
+                next_instr--;
+                if (_Py_Specialize_LoadMethod(owner, next_instr, name, cache) < 0) {
+                    goto error;
+                }
+                DISPATCH();
+            }
+            else {
+                STAT_INC(LOAD_METHOD, deferred);
+                cache->adaptive.counter--;
+                oparg = cache->adaptive.original_oparg;
+                STAT_DEC(LOAD_METHOD, unquickened);
+                JUMP_TO_INSTRUCTION(LOAD_METHOD);
+            }            
+        }
+
+        case TARGET(LOAD_METHOD_CACHED): {
```

Comment:
Since almost all builtin instances have no dict (`tp_dictoffset == 0`) and almost all Python instances do have a dict (`tp_dictoffset != 0`), which is usually split. it might be worth splitting this into two specialized forms.

`LOAD_METHOD_SPLIT_KEYS` and `LOAD_METHOD_NO_DICT`. The `NO_DICT` form would only two read two cache slots.

Reply:
I'd considered that in the original issue on faster-cpython. But IMO the maintenance burden would be much greater, and I don't know how much that would speed up pyperformance (or even microbenchmarks for that matter). My unvalidated hunch is not by much.